utils/service_discovery/sd_docker_backend.py

```diff
@@ -32,49 +32,100 @@ def __init__(self, agentConfig):
+        tpl_parts = tpl_var.split('_')
+
+        # a specifier was given
+        if len(tpl_parts) > 1:
+            networks = c_inspect.get('NetworkSettings', {}).get('Networks') or {}
+            ip_dict = {}
+            for net_name, net_desc in networks.iteritems():
+                ip = net_desc.get('IPAddress')
+                if ip:
+                    ip_dict[net_name] = ip
+            ip_addr = self._extract_ip_from_networks(ip_dict, tpl_var)
+            if ip_addr:
+                return ip_addr
+
+        # try to get the bridge IP address
+        log.debug("No network found for container %s (%s), trying with IPAddress field" % (c_id[:12], c_img))
+        ip_addr = c_inspect.get('NetworkSettings', {}).get('IPAddress')
+        if ip_addr:
+            return ip_addr
 
-        return ip_addr
+        else:
+            if is_k8s():
+                # kubernetes case
+                log.debug("Couldn't find the IP address for container %s (%s), "
+                          "using the kubernetes way." % (c_id[:12], c_img))
+                pod_list = self.kubeutil.retrieve_pods_list().get('items', [])
+                for pod in pod_list:
+                    pod_ip = pod.get('status', {}).get('podIP')
+                    if pod_ip is None:
+                        continue
+                    else:
+                        c_statuses = pod.get('status', {}).get('containerStatuses', [])
+                        for status in c_statuses:
+                            # compare the container id with those of containers in the current pod
+                            if c_id == status.get('containerID', '').split('//')[-1]:
+                                return pod_ip
+
+        log.error("No IP address was found for container %s (%s)" % (c_id[:12], c_img))
+        return None
+
+    def _extract_ip_from_networks(self, ip_dict, tpl_var):
+        """Extract a single IP from a dictionary made of network names and IPs."""
+        if not ip_dict:
+            return None
+        tpl_parts = tpl_var.split('_')
+
+        # no specifier, try to pick the bridge key, falls back to the value of the last key
+        if len(tpl_parts) < 2:
+            log.debug("No key was passed for template variable %s." % tpl_var)
+            if 'bridge' in ip_dict:
+                log.debug("Using the bridge network instead.")
+                return ip_dict['bridge']
+            else:
+                last_key = sorted(ip_dict.iterkeys())[-1]
+                log.debug("Trying with the last key: '%s' instead." % last_key)
+                return ip_dict[last_key]
+        else:
+            res = ip_dict.get(tpl_parts[-1])
+            if res is None:
+                log.warning("The key passed for template variable %s was not found." % tpl_var)
+                if 'bridge' in ip_dict:
+                    log.warning("Using the bridge network instead.")
```

Comment:
If no key is passed because there's no need for it, there is probably no need for a warning whereas if a key was explicitly passed but not found it definitely requires a warning. But I'm tempted to log both as warning. Better be a little spammy than too obscure.